BSP/Simulator_V2/Src/Simulator.c

```diff
@@ -88,46 +113,90 @@ static void readInputFile(void) {
             tail->next = temp;
             tail = tail->next;
         }
-        // iterate through a single simulator state
-        for (cJSON *field = state->child; field != NULL; field = field->next) {
-            if (strcmp(field->string, "time") == 0) {
-                tail->time = field->valueint;
-            } else if (strcmp(field->string, "adcHigh") == 0) {
-                tail->adcHigh = field->valueint;
-            } else if (strcmp(field->string, "adcLow") == 0) {
-                tail->adcLow = field->valueint;
-            } else if (strcmp(field->string, "voltages") == 0) {
-                // get all of the voltages
-                int i = 0;
-                for (cJSON *voltage = field->child; voltage != NULL && i < NUM_BATTERY_MODULES; i++, voltage = voltage->next) {
-                    tail->voltages[i] = voltage->valueint;
-                }
-                if (i != NUM_BATTERY_MODULES) {
-                    printf("Error: incorrect number of voltages! Expected %d, but got %d\n", NUM_BATTERY_MODULES, i);
-                    exit(-1);
-                }
-            } else if (strcmp(field->string, "temperatures") == 0) {
-                // get all of the temperatures
-                int i = 0;
-                for (cJSON *temperature = field->child; 
-                    temperature != NULL && i < (NUM_BATTERY_MODULES * NUM_TEMP_SENSORS_PER_MOD); 
-                    i++, temperature = temperature->next) {
-
-                    tail->temperatures[i] = temperature->valueint;
-                }
-                if (i != (NUM_BATTERY_MODULES * NUM_TEMP_SENSORS_PER_MOD)) {
-                    printf("Error: incorrect number of temperatures! Expected %d, but got %d\n", (NUM_BATTERY_MODULES * NUM_TEMP_SENSORS_PER_MOD), i);
+        char buffer[50];
+        sprintf(buffer, "\nExecuting state #%d...\n", stateCount);
+        Simulator_log(buffer);
+        // Try and get every field we should have in each state.
+        // If null, it doesn't exist.
+        cJSON* time = cJSON_GetObjectItem(state, "time");
+        cJSON* adcHigh = cJSON_GetObjectItem(state, "adcHigh");
+        cJSON* adcLow = cJSON_GetObjectItem(state, "adcLow");
+        // voltageArray should have 31 elements.
+        cJSON* voltageArray = cJSON_GetObjectItem(state, "voltages");
+        // temperatureArray should have 31*2 elements.
+        cJSON* temperatureArray = cJSON_GetObjectItem(state, "temperatures");
+        cJSON* current = cJSON_GetObjectItem(state, "current");
+        cJSON* charge = cJSON_GetObjectItem(state, "charge");
+        cJSON* canList = cJSON_GetObjectItem(state, "can");
+    
+        // Check which state fields aren't present in the current sim state JSON.
+        // Lots of nasty if/else's here, not much way around this.
+        if (!time) {
+            printf("Current simulator state does not have a specified time value. Using -1...\n");
```

Comment:
Might want to make these use the simulator log, but I don't feel strongly about it. We can always change it later if these get annoying

Reply:
I thought about that but decided on printf because you don't immediately see the output of the sim without opening the log file, and thought printfs() were a little more informative in the moment of running the sim.